clients/lsp-rust.el

```diff
@@ -860,6 +866,45 @@ them with `crate` or the crate name they refer to."
        (lambda (_) (concat "*" label "*")))
       (setq lsp-rust-analyzer--last-runnable runnable))))
 
+(defun lsp-rust-analyzer-debug (runnable)
+  "Select and run a runnable action."
+  (interactive (list (lsp-rust-analyzer--select-runnable)))
+  (unless (featurep 'dap-cpptools)
+    (user-error "You must require `dap-cpptools'."))
+  (-let (((&rust-analyzer:Runnable
+           :args (&rust-analyzer:RunnableArgs :cargo-args :workspace-root? :executable-args)
+           :label) runnable))
+    (cl-case (aref cargo-args 0)
+      ("run" (aset cargo-args 0 "build"))
+      ("test" (when (-contains? (append cargo-args ()) "--no-run")
+                (cl-callf append cargo-args (list "--no-run")))))
+    (->> (append (list (executable-find "cargo"))
+                 cargo-args
+                 (list "--message-format=json"))
+         (s-join " ")
+         (shell-command-to-string)
+         (s-lines)
+         (-keep (lambda (s)
+                  (condition-case _
+                      (-let* ((json-object-type 'plist)
+                              ((msg &as &plist :reason :executable) (json-read-from-string s)))
+                        (when (and executable (string= "compiler-artifact" reason))
+                          executable))
+                    (error))))
+         (funcall (lambda (artifact-spec)
+                    (pcase artifact-spec
+                      (`() (user-error "No compilation artifacts or reading the "))
+                      (`(,spec) spec)
+                      (_ (user-error "Multiple compilation artifacts are not supported.")))))
+         (list :type "cppdbg"
```

Comment:
In the future, we should consider supporting other debuggers. This isn't a critique though, as it is probably out of scope here.

Reply:
Yes, support for all of the possible rust-friendly debug adapters is with lower priority.